src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java

```diff
@@ -133,6 +136,7 @@
   @Nullable private final RemoteExecutionClient remoteExecutor;
   private final ImmutableSet<PathFragment> filesToDownload;
   @Nullable private final Path captureCorruptedOutputsDir;
+  private final ConcurrentHashMap<Object, MerkleTree> merkleTreeCache = new ConcurrentHashMap<>();
```

Comment:
Shall I use soft references to the `MerkleTree`s or some other kind of map to reduce the memory footprint?

Reply:
I got out of memory using `ConcurrentHashMap`. Switching to `Cache<Object, MerkleTree> merkleTreeCache = CacheBuilder.newBuilder().softValues().build()` works for me. I'll fix that and all the other comments (and test errors) next week.